From 5afdf5536517c60a886b259426063bdb88ecf5c8 Mon Sep 17 00:00:00 2001
From: Alex Van Damme <atv@google.com>
Date: Fri, 16 Jan 2026 17:50:34 +0000
Subject: [PATCH 4/4] Add custom CoralNPU CSRs and update MVENDORID/MARCHID

---
 riscv/csr_init.cc | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/riscv/csr_init.cc b/riscv/csr_init.cc
index 5dcbb641..60a71477 100644
--- a/riscv/csr_init.cc
+++ b/riscv/csr_init.cc
@@ -324,9 +324,9 @@ void state_t::csr_init(processor_t* const proc, reg_t max_isa)
 
   add_ext_csr(EXT_ZKR, CSR_SEED, std::make_shared<seed_csr_t>(proc, CSR_SEED));
 
-  add_csr(CSR_MARCHID, std::make_shared<const_csr_t>(proc, CSR_MARCHID, 5));
+  add_csr(CSR_MARCHID, std::make_shared<const_csr_t>(proc, CSR_MARCHID, 0));
   add_csr(CSR_MIMPID, std::make_shared<const_csr_t>(proc, CSR_MIMPID, 0));
-  add_csr(CSR_MVENDORID, std::make_shared<const_csr_t>(proc, CSR_MVENDORID, 0));
+  add_csr(CSR_MVENDORID, std::make_shared<const_csr_t>(proc, CSR_MVENDORID, 0x426));
   add_csr(CSR_MHARTID, std::make_shared<const_csr_t>(proc, CSR_MHARTID, proc->get_id()));
   add_csr(CSR_MCONFIGPTR, std::make_shared<const_csr_t>(proc, CSR_MCONFIGPTR, 0));
 
@@ -337,6 +337,16 @@ void state_t::csr_init(processor_t* const proc, reg_t max_isa)
   add_csr(0xFCC, std::make_shared<basic_csr_t>(proc, 0xFCC, 0)); // KSCM2
   add_csr(0xFD0, std::make_shared<basic_csr_t>(proc, 0xFD0, 0)); // KSCM3
   add_csr(0xFD4, std::make_shared<basic_csr_t>(proc, 0xFD4, 0)); // KSCM4
+  add_csr(0x7C0, std::make_shared<basic_csr_t>(proc, 0x7C0, 0)); // MCONTEXT0
+  add_csr(0x7C1, std::make_shared<basic_csr_t>(proc, 0x7C1, 0)); // MCONTEXT1
+  add_csr(0x7C2, std::make_shared<basic_csr_t>(proc, 0x7C2, 0)); // MCONTEXT2
+  add_csr(0x7C3, std::make_shared<basic_csr_t>(proc, 0x7C3, 0)); // MCONTEXT3
+  add_csr(0x7C4, std::make_shared<basic_csr_t>(proc, 0x7C4, 0)); // MCONTEXT4
+  add_csr(0x7C5, std::make_shared<basic_csr_t>(proc, 0x7C5, 0)); // MCONTEXT5
+  add_csr(0x7C6, std::make_shared<basic_csr_t>(proc, 0x7C6, 0)); // MCONTEXT6
+  add_csr(0x7C7, std::make_shared<basic_csr_t>(proc, 0x7C7, 0)); // MCONTEXT7
+  add_csr(0x7E0, std::make_shared<basic_csr_t>(proc, 0x7E0, 0)); // MPC
+  add_csr(0x7E1, std::make_shared<basic_csr_t>(proc, 0x7E1, 0)); // MSP
 
   const reg_t senvcfg_mask = (proc->extension_enabled(EXT_ZICBOM) ? SENVCFG_CBCFE | SENVCFG_CBIE : 0) |
                             (proc->extension_enabled(EXT_ZICBOZ) ? SENVCFG_CBZE : 0) |
-- 
2.43.7

